cli.js

```diff
@@ -1,20 +1,128 @@
+    compiler.javaPath = getNativeImagePath();
+  }
+
+  compiler.run((exitCode) => {
+    process.exit(exitCode);
+  });
+} else {
+  if (compilerFlags.help === true) {
+    console.log('Sample usage: --compilation_level (-O) VAL --externs VAL --js VAL --js_output_file VAL --warning_level (-W) [QUIET | DEFAULT | VERBOSE]');
+    console.log('See https://github.com/google/closure-compiler/wiki/Flags-and-Options for the full list of flags');
+    process.exit(0);
+  }
+
+  if (compilerFlags.version === true) {
+    const {version} = require('./package.json');
+    console.log(`Version: ${version}`);
+    process.exit(0);
+  }
+
+  let inputFilePromises = [];
+  let waitOnStdIn = true;
+  if (compilerFlags.js) {
+    waitOnStdIn = false;
+    if (!Array.isArray(compilerFlags.js)) {
+      compilerFlags.js = [compilerFlags.js];
+    }
+    inputFilePromises = compilerFlags.js.map(path =>
+        new Promise((resolve, reject) =>
+            fs.readFile(path, 'utf8', (err, src) => err ? reject(err) : resolve({src, path})))
+        .catch(e => console.error(e)));
+
+    delete compilerFlags.js;
+  }
+  let externFilePromises = [];
+  if (compilerFlags.externs) {
+    if (!Array.isArray(compilerFlags.externs)) {
+      compilerFlags.externs = [compilerFlags.externs];
+    }
+    externFilePromises = compilerFlags.externs.map(path =>
+        new Promise((resolve, reject) =>
+            fs.readFile(path, 'utf8', (err, src) => err ? reject(err) : resolve({src, path})))
+        .catch(e => console.error(e)));
+
+    delete compilerFlags.externs;
+  }
+
+  const inputsReadPromise = inputFilePromises.length > 0 ? Promise.all(inputFilePromises) : Promise.resolve([]);
+  const externsReadPromise = externFilePromises.length > 0 ? Promise.all(externFilePromises) : Promise.resolve([]);
+
+  Promise.all([inputsReadPromise, externsReadPromise])
+    .then(([inputFiles, externs]) => {
+      if (externs.length > 0) {
+        compilerFlags.externs = externs;
+      } else {
+        delete compilerFlags.externs;
+      }
+
+      if (!waitOnStdIn) {
+        return inputFiles;
+      } else {
+        return new Promise(resolve => {
+          let stdInData = '';
+          process.stdin.on('readable', () => {
+            const chunk = process.stdin.read();
+            if (chunk !== null) {
+              stdInData += chunk;
+            }
+          });
+          process.stdin.on('end', () => {
+            if (stdInData.length > 0) {
+              inputFiles.push({
+                path: 'stdin',
+                src: stdInData
+              });
+            }
+            resolve(inputFiles);
+          });
+        });
+      }
+    })
+    .then(inputFiles => {
+      const Compiler = require('./lib/node/closure-compiler-js');
+      const compiler = new Compiler(compilerFlags);
+      compiler.run(inputFiles, (exitCode, compiledFiles, errors) => {
+        if (errors && errors.length > 0) {
+          console.error(errors);
```

Comment:
Do we know that errors.length !== 0 -> exitCode !== 0?

Reply:
I wasn't willing to bet on that.